### PR TITLE
Add non-owning SampleView

### PR DESCRIPTION
**Pull request by klecki**
## Category: **New feature**  

Replacing #3700

## Description: 
Add non-owning SampleView

The purpose is to have Tensor wrapper that is similar
to already existing DALI vocabulary types representing
Tensor:
* it is non-owning view to the data
* it allows to access the pointer, the shape and the type id
* can be used as argument to view the same way as Tensor
but as opposed to Tensor doesn't allow to adjust the allocation.

The target of this type is to use it as the return
type of the TensorBatch::operator[] - using static type
is not possible due to the Tensor[List/Vector] not being
statically typed. This should allow for proper encapsulation
of the batch object - SampleView allows access to the data
but does not allow to adjust the allocation or changing
the metadata through the view - as oposed to the currently
used Tensor.

Add conversion from SampleView to statically typed TensorView
using view<T, ndim> template function.

### To consider
Options for follow-up:
* shared_ptr as member to have co-ownership instead of non-owning type.

### Affected modules and functionalities:
views

### Key points relevant for the review:
Proper encapsulation?

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/data/sample_view.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    return data_;
+  }
+
+  const void *data() const {
+    return data_;
+  }
+
+  const void *cdata() const {
+    return data_;
+  }
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:53]: If this type is to stay, I'd suggest that the API be consistent with existing buffer types.

File: dali/pipeline/data/sample_view.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  SampleView &operator=(const SampleView &) = default;
+
+  SampleView(SampleView &&other) {
+    data_ = other.data_;
+    other.data_ = nullptr;
+    shape_ = std::move(other.shape_);
+    other.shape_ = {};
+    type_id_ = other.type_id_;
+    other.type_id_ = DALI_NO_TYPE;
+  }
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:79]: This doesn't make much sense if we allow copying.

File: dali/pipeline/data/sample_view.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    if (this != &other) {
+      data_ = other.data_;
+      other.data_ = nullptr;
+      shape_ = std::move(other.shape_);
+      other.shape_ = {};
+      type_id_ = other.type_id_;
+      other.type_id_ = DALI_NO_TYPE;
+    }
+    return *this;
+  }
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:91]: Likewise.

File: dali/pipeline/data/sample_view.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/data/sample_view.h]: I would really like to introduce some more std-like naming here, but maybe as a next step.

File: dali/pipeline/data/sample_view.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  SampleView() = default;
+
+  SampleView(const SampleView &) = default;
+  SampleView &operator=(const SampleView &) = default;
+
+  SampleView(SampleView &&other) {
+    data_ = other.data_;
+    other.data_ = nullptr;
+    shape_ = std::move(other.shape_);
+    other.shape_ = {};
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:116]:
```suggestion
    other.shape_ = { 0 };
```

File: dali/pipeline/data/sample_view.h [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    type_id_ = other.type_id_;
+    other.type_id_ = DALI_NO_TYPE;
+  }
+
+  SampleView &operator=(SampleView &&other) {
+    if (this != &other) {
+      data_ = other.data_;
+      other.data_ = nullptr;
+      shape_ = std::move(other.shape_);
+      other.shape_ = {};
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:126]:
```suggestion
      other.shape_ = { 0 };
```

File: dali/pipeline/data/sample_view.h [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  SampleView &operator=(const SampleView &) = default;
+
+  SampleView(SampleView &&other) {
+    data_ = other.data_;
+    other.data_ = nullptr;
+    shape_ = std::move(other.shape_);
+    other.shape_ = {};
+    type_id_ = other.type_id_;
+    other.type_id_ = DALI_NO_TYPE;
+  }
```
Comment (mzient) [inline dali/pipeline/data/sample_view.h:119]: Reuse implementation?
```suggestion
  SampleView(SampleView &&other) {
    *this = std::move(other);
  }
```

File: dali/pipeline/data/sample_view.h [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+   */
+  SampleView(void *data, const TensorShape<> shape, DALIDataType type_id)
+      : data_(data), shape_(std::move(shape)), type_id_(type_id) {}
+
+ private:
+  // TODO(klecki): The view is introduced with no co-owning pointer, it will be evaluated
+  // if the usage of shared_ptr is possbile and adjusted if necessary.
+  // Using shared_ptr might allow for sample exchange between two batches using operator[]
+  void *data_ = nullptr;
+  TensorShape<> shape_ = {};
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/data/sample_view.h:151]: done, although I think it's misleading in other ways. We need a better empty tensor shape default

File: dali/pipeline/data/sample_view.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/data/sample_view.h:126]: done

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/sample_view.h [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+class SampleView {
+ public:
+  /**
+   * @name Get the underlying pointer to data
+   */
+  // @{
+  /**
+   * @brief Return an un-typed pointer to the underlying storage.
+   * A valid type must be set prior to calling this method by calling
+   * the non-const version of the method, or calling 'set_type'.
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/data/sample_view.h:46]: I got carried away with the copy-paste.

**Comment by klecki** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by stiepan — APPROVED** [review]

**Pull request merged**